reconcile/ocm_clusters.py

```diff
@@ -1,76 +1,248 @@
+
+    if not desired_spec.elb_fqdn:
+        root_updates[
+            ocmmod.SPEC_ATTR_ELBFQDN
+        ] = f"elb.apps.{cluster}.{current_spec.domain}"
+
+    updates: dict[str, Any] = {}
+    updates[ocmmod.SPEC_ATTR_PATH] = desired_spec.path
+    updates["root"] = root_updates
+    updates["spec"] = ocm_spec_updates
+
+    return updates, error
+
+
+def get_cluster_ocm_update_spec(
+    ocm: ocmmod.OCM, cluster: str, current_spec: OCMSpec, desired_spec: OCMSpec
+) -> Tuple[dict[str, Any], bool]:
+    """Get cluster updates to request to OCM api
+
+    Parameters:
+        ocm (OCM): ocm implementation for an ocm product (osd, rosa)
+        cluster (str): cluster name
+        current_spec (OCMSpec): Cluster spec retreived from OCM api
+        desired_spec (OCMSpec): Cluster spec retreived from App-Interface
+
+    Returns:
+        updates (dict): Updates to request to OCM api
+        error (bool): If errors detected due to spec not allowed updates.
+    """
+
+    impl = ocmmod.OCM_PRODUCTS_IMPL[current_spec.spec.product]
 
     error = False
-    if current_spec["network"] != desired_spec["network"]:
+    if not desired_spec.network.type:
+        desired_spec.network.type = "OpenShiftSDN"
+
+    if current_spec.network != desired_spec.network:
         error = True
-        logging.error(f"[{cluster_name}] invalid update: network")
-    current_spec_spec = current_spec["spec"]
-    desired_spec_spec = desired_spec["spec"]
-    updated = {
-        k: desired_spec_spec[k]
-        for k in desired_spec_spec
-        if current_spec_spec.get(k) != desired_spec_spec[k]
+        logging.error(f"[{cluster}] invalid update: network")
+
+    # Convert ocm specs to dicts, removing null values and excluded attributes
+    current_ocm_spec = {
+        k: v
+        for k, v in current_spec.spec.dict().items()
+        if v is not None and k not in impl.EXCLUDED_SPEC_FIELDS
     }
 
-    # we only need deleted to check if a field removal is valid
-    # we really want to check updated + deleted, and since
-    # we have no further use for deleted -
-    deleted = {
-        k: current_spec_spec[k] for k in current_spec_spec if k not in desired_spec_spec
+    desired_ocm_spec = {
+        k: v
+        for k, v in desired_spec.spec.dict().items()
+        if v is not None and k not in impl.EXCLUDED_SPEC_FIELDS
     }
-    diffs = deleted
-    diffs.update(updated)
 
-    invalid_fields = set(diffs.keys()) - ALLOWED_SPEC_UPDATE_FIELDS
-    if invalid_fields:
+    # Updated attributes in app-interface
+    updated_attrs = {
+        k: v for k, v in desired_ocm_spec.items() if current_ocm_spec.get(k) != v
+    }
+
+    # Removed attributes in app-interface
+    deleted_attrs = {
+        k: v
+        for k, v in current_ocm_spec.items()
+        if k not in desired_ocm_spec and v is not None
+    }
+
+    diffs = deleted_attrs
+    diffs.update(updated_attrs)
+
+    not_allowed_updates = set(diffs) - impl.ALLOWED_SPEC_UPDATE_FIELDS
```

Comment:
can you handle the forbidden network update with the ALLOWED_SPEC_UPDATE_FIELDS approach?

Reply:
The model used is a bit misleading, the network is not part of the cluster spec but the ocm spec. I've changed the code to compare the OCM spec instead of the Cluster spec only, so the network and the cluster spec are joined in a new dict before doing the comparison. 

Thanks, I think this is more elegant now.